## Imports:
import os
import pandas as pd
from pathlib import Path
from typing import List

# check if param --config-file is set 
configfile: "config/config.yaml" 

# Check if verbose is set
verbose = config["verbose"]

# Check for all paths in config file if they are set or exist and set them to default if not
illumina_samples = Path(config["illumina_samples"])
nanopore_sampels = Path(config["nanopore_samples"])
primer_illumina = Path(config["primer_illumina"])
primer_nanopore = Path(config["primer_nanopore"])
reference_genome = Path(config["ref"])

if not (illumina_samples.is_dir() or nanopore_sampels.is_dir()):
    raise IOError("Illumina and Nanopore samples directory does not exist! \
                Make sure, one of the folders configures in the `cconfig.yaml` \
                file exists and contains samples!")

results_dir = Path(config["results"])
if results_dir == "" or results_dir == None:
    if verbose:
        print("== Result dir not set in config - will be set to default path `./results`!")
    results_dir = Path("./results")
results_dir.mkdir(parents=True, exist_ok=True)

envs_dir = Path(config["envs_dir"])
if envs_dir == "" or envs_dir == None:
    if verbose:
        print("== Env dir not set in config - will be set to default path `../envs`!")
    envs_dir = Path("../envs")

# Define subdirectories
benchmark_dir = Path(results_dir / "benchmark")
benchmark_dir.mkdir(parents=True, exist_ok=True)
log_dir = Path(results_dir / "log")
log_dir.mkdir(parents=True, exist_ok=True)


def dir_to_sample_df(dir: Path, pairwise: bool) -> pd.DataFrame:
# Load illumina samples by scanning the directory for fastq files and use the sample name as index column and the file path as values in the dataframe, differentiating between R1 and R2
    list_of_files: List[Path] = [f for f in dir.glob("*.fastq.gz") if f.is_file()]
    sample_df: pd.DataFrame = pd.DataFrame()
    if pairwise:
        sample_df["R1"] = [str(f) for f in list_of_files if ".R1." in str(f)]
        sample_df["R2"] = [str(f) for f in list_of_files if ".R2." in str(f)]
    else:
        sample_df["R1"] = [str(f) for f in list_of_files]
    sample_df["sample"] = [Path(f).stem.split(".")[0] for f in sample_df["R1"]]
    return sample_df.set_index("sample")

illumina_samples_df = dir_to_sample_df(illumina_samples, pairwise=True)
print(illumina_samples_df)
nanopore_samples_df = dir_to_sample_df(nanopore_sampels, pairwise=False)
print(nanopore_samples_df)

# Include rule files:
include: os.path.join("rules", "qc.smk")
include: os.path.join("rules", "mapping.smk")
include: os.path.join("rules", "primer_clipping.smk")
include: os.path.join("rules", "variant_calling.smk")

# Rule all
# TODO: adjust to use wildcards for all samples and all primer combinations in EVERY rule, so no hard-coded stuff
# This will also lead to a nice DAG overview of the whole pipeline
rule all:
    input: 
        "output/qc/clean_reads.R1.fastq.gz",
        "output/qc/nanoplot/raw/NanoPlot-report.html",
        "output/qc/clean_reads_nanopore.fastq.gz",
        "output/mapping/minimap2-illumina.sam",
        "output/mapping/minimap2-nanopore.sam",
        expand("output/mapping/{sample}.sorted.bam", sample=config["samples"]),
        expand("output/mapping/{sample}.sorted.bam.bai", sample=config["samples"]),
        expand("results/primer_scheme/converted-{tech}.bedpe", tech=config["tech"]),
        expand("results/primer_scheme/corrected-{tech}.bedpe", tech=config["tech"]),
        expand("results/primer_clipping/{tech}_clipped.bam", tech=config["tech"]),
        
# TODO: Python script to generate the Benchmarking-Plots from the output of the pipeline. Call it in rule all so everything gets executed in one go